lib/vra/resource.rb

```diff
@@ -0,0 +1,174 @@
+#
+# Author:: Chef Partner Engineering (<partnereng@chef.io>)
+# Copyright:: Copyright (c) 2015 Chef Software, Inc.
+# License:: Apache License, Version 2.0
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+module Vra
+  # rubocop:disable ClassLength
+  class Resource
+    attr_reader :id, :resource_data
+
+    def initialize(client, opts)
+      @client           = client
+      @id               = opts[:id]
+      @resource_data    = opts[:data]
+      @resource_actions = []
+
+      if @id.nil? && @resource_data.nil?
+        raise ArgumentError, 'must supply an id or a resource data hash'
+      end
+
+      if ! @id.nil? && ! @resource_data.nil?
+        raise ArgumentError, 'must supply an id OR a resource data hash, not both'
+      end
+
+      if @resource_data.nil?
+        fetch_resource_data
+      else
+        @id = @resource_data['id']
+      end
+    end
+
+    def fetch_resource_data
+      @resource_data = JSON.load(@client.http_get!("/catalog-service/api/consumer/resources/#{@id}"))
+    rescue Vra::Exception::HTTPNotFound
+      raise Vra::Exception::NotFound, 'resource ID #{@id} does not exist'
+    end
+
+    def name
+      @resource_data['name']
+    end
+
+    def status
+      @resource_data['status']
+    end
+
+    def vm?
+      @resource_data['resourceTypeRef']['id'] == 'Infrastructure.Virtual'
+    end
+
+    def tenant_id
+      @resource_data['organization']['tenantRef']
+    end
+
+    def tenant_name
+      @resource_data['organization']['tenantLabel']
+    end
+
+    def subtenant_id
+      @resource_data['organization']['subtenantRef']
+    end
+
+    def subtenant_name
+      @resource_data['organization']['subtenantLabel']
+    end
+
+    def catalog_id
+      @resource_data['catalogItem']['id']
+    end
+
+    def catalog_name
+      @resource_data['catalogItem']['label']
+    end
+
+    def network_interfaces
+      return unless vm?
+
+      nics = []
```

Comment:
Yeah, I could totally refactor lines 94-103 to just map/inject and return the outcome of that operation rather than building a new array.  I will be MOAR RUBYISH.